src/dft/src/ScanReplace.cpp

```diff
@@ -0,0 +1,426 @@
+  for (odb::dbLib* lib : db_->getLibs()) {
+    for (odb::dbMaster* master : lib->getMasters()) {
+      // We only care about sequential cells in DFT
+      if (!master->isSequential()) {
+        continue;
+      }
+
+      sta::Cell* master_cell = db_network->dbToSta(master);
+      sta::LibertyCell* liberty_cell = db_network->libertyCell(master_cell);
+
+      if (!liberty_cell) {
+        // Without info about the cell we can't do any replacement
+        continue;
+      }
+
+      // If the lib cell is marked as dontUse, then skip it
+      if (liberty_cell->dontUse()) {
+        continue;
+      }
+
+      // We skip clock gating lib cells
+      if (liberty_cell->isClockGate()) {
+        continue;
+      }
+
+      if (IsScanCell(liberty_cell)) {
+        available_scan_lib_cells_.insert(liberty_cell);
+      } else if (CanBeScanReplace(master, liberty_cell)) {
+        non_scan_cells.push_back(liberty_cell);
+      }
+    }
+  }
+
+  // Let's find what are the scan equivalent cells for each non scan
+  std::unordered_map<sta::LibertyCell*,
+                     std::vector<std::unique_ptr<ScanCandidate>>>
+      scan_candidates;
+  for (sta::LibertyCell* non_scan_cell : non_scan_cells) {
+    for (sta::LibertyCell* scan_cell : available_scan_lib_cells_) {
+      std::unordered_map<std::string, std::string> port_mapping;
+      if (IsScanEquivalent(non_scan_cell, scan_cell, port_mapping)) {
+        scan_candidates[non_scan_cell].push_back(
+            std::make_unique<ScanCandidate>(scan_cell, port_mapping));
+      }
+    }
+
+    // Check if this non_scan_cell have scan equivalents
+    auto found = scan_candidates.find(non_scan_cell);
+    if (found == scan_candidates.end()) {
+      logger_->warn(utl::DFT,
+                    1,
+                    "Lib cell '{:s}' doesn't have an scan equivalent",
+                    non_scan_cell->name());
+    }
+  }
+
+  // Populate non_scan_to_scan_cells to map non scan to scan cells with just the
+  // best equivalent scan cell
+  for (auto& candidates : scan_candidates) {
+    sta::LibertyCell* non_scan_cell = candidates.first;
+    non_scan_to_scan_lib_cells_.insert(
+        {non_scan_cell, SelectBestScanCell(non_scan_cell, candidates.second)});
+  }
+}
+
+ScanReplace::ScanReplace(odb::dbDatabase* db,
+                         sta::dbSta* sta,
+                         utl::Logger* logger)
+    : db_(db), sta_(sta), logger_(logger)
+{
+  db_network_ = sta->getDbNetwork();
+}
+
+void ScanReplace::scanReplace()
+{
+  odb::dbChip* chip = db_->getChip();
+  scanReplace(chip->getBlock());
+}
+
+// Recursive function that iterates over a block (and the blocks inside this
+// one) replacing the cells with scan equivalent
+void ScanReplace::scanReplace(odb::dbBlock* block)
+{
+  odb::dbNet* ground_net = utils::FindGroundNet(db_network_, block);
+
+  for (odb::dbInst* inst : block->getInsts()) {
+    if (inst->isDoNotTouch()) {
```

Comment:
You also need to look at whether the iterms connect to dont touch nets as you are going to modify them by disconnect/connect

Reply:
Thanks!, I added an additional check